smqtk_iqr/iqr/iqr_session.py

```diff
@@ -300,25 +316,72 @@ def refine(self) -> None:
 
         """
         with self.lock:
-            # combine pos/neg adjudications + added external data descriptors
+            # Combine pos/neg adjudications + added external data descriptors
             pos = [desc.vector() for desc in (self.positive_descriptors |
                                               self.external_positive_descriptors)]
-
             neg = [desc.vector() for desc in (self.negative_descriptors |
                                               self.external_negative_descriptors)]
 
             if not pos:
                 raise RuntimeError("Did not find at least one positive "
                                    "adjudication.")
 
-            self._log.debug("Ranking working set with %d pos and %d neg total "
-                            "examples.", len(pos), len(neg))
+            # Get working set descriptors
             pool_uids, pool_de = zip(*self.working_set.items())
             pool = [de.vector() for de in pool_de]
+
+            # Auto-select negative examples if none are given
+            if not neg:
+                neg_autoselect = set()
+                self._log.info(f"Auto-selecting negative examples. \
+                              ({self.autoneg_select_ratio} per positive)")
+                # For each positive example, find the farthest descriptor
+                # from it to use as a negative example
+                for p in pos:
+                    # Compute distance between current positive example and
+                    # each of the working set descriptors
+                    distances = []
+                    for desc_v in pool:
+                        distances.append(self.distance_metric(p, np.array(desc_v)))
+
+                    np_distances: np.array = np.array(distances)
+
+                    # Scan vector for max distance index
+                    # - Allow variable number of maximally distance descriptors to
+                    #   be picked per positive.
+                    # track most distance neighbors
+                    m_set = {}
+                    # track smallest distance of most distant neighbors
+                    m_val = -float('inf')
+                    for i in range(np_distances.size):
+                        if np_distances[i] > m_val:
+                            m_set[np_distances[i]] = i
+
+                            if len(m_set) > self.autoneg_select_ratio:
+                                if m_val in m_set:
+                                    del m_set[m_val]
+                            m_val = min(m_set)
+
+                    for i in m_set.values():
+                        neg_autoselect.add(pool_de[i])
+
+                # Remove any positive examples from auto-selected results
+                neg_autoselect.difference_update((self.positive_descriptors |
+                                                  self.external_positive_descriptors))
```

Comment:
There is technically a remote possibility that this will remove all auto-selected elements. We should raise a documented exception here in such a case. We'll likely never see it, but it's easy to be thorough in this case.